server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1383,14 +1528,26 @@ protected Entitlement addOrUpdateEntitlement(Consumer consumer, Pool pool,
      * @param consumer
      * @param pool
      */
-    private void checkBonusPoolQuantities(Pool pool, Entitlement e) {
-        for (Pool derivedPool :
-                lookupOversubscribedBySubscriptionId(pool.getSubscriptionId(), e)) {
-            if (!derivedPool.getId().equals(pool.getId()) &&
-                derivedPool.getQuantity() != -1) {
-                deleteExcessEntitlements(derivedPool);
+    private void checkBonusPoolQuantities(Map<String, PoolQuantity> poolQuantities,
+            Map<String, Entitlement> entitlements) {
+        Set<String> excludePoolIds = new HashSet<String>();
+        Map<String, Entitlement> subEntitlementMap = new HashMap<String, Entitlement>();
+        for (Entry<String, PoolQuantity> entry : poolQuantities.entrySet()) {
+            Pool pool = entry.getValue().getPool();
+            subEntitlementMap.put(pool.getSubscriptionId(),
+                    entitlements.get(entry.getKey()));
+            excludePoolIds.add(pool.getId());
+        }
+
+        List<Pool> overConsumedPools = poolCurator.lookupOversubscribedBySubscriptionIds(subEntitlementMap);
+
+        List<Pool> derivedPools = new ArrayList<Pool>();
+        for (Pool pool : overConsumedPools) {
+            if (!excludePoolIds.contains(pool.getId()) && pool.getQuantity() != -1) {
```

Comment:
Micro-optimization: It's slightly faster to flip the conditions around here, as the quantity check is a simple value comparison vs. performing a list iteration.